libs/vkd3d/command.c

```diff
@@ -3474,87 +3474,231 @@ static HRESULT d3d12_command_list_allocate_transfer_buffer(struct d3d12_command_
 
-    VK_CALL(vkCmdCopyBufferToImage(list->vk_command_buffer,
-            transfer_buffer.vk_buffer, dst_resource->u.vk_image,
-            VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &buffer_image_copy));
+        if (region->extent.width == extent.width && region->extent.height == extent.height)
+            vk_image_barriers[0].oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+
+        vk_image_barriers[1].oldLayout = src_layout;
+        vk_image_barriers[1].newLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+        vk_image_barriers[1].dstAccessMask = VK_ACCESS_SHADER_READ_BIT;
+        vk_image_barriers[1].image = src_resource->u.vk_image;
+        vk_image_barriers[1].subresourceRange = vk_subresource_range_from_layers(&region->srcSubresource);
+
+        VK_CALL(vkCmdPipelineBarrier(list->vk_command_buffer,
+                VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
+                VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT |
+                VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT |
+                VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT |
+                VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+                0, 0, NULL, 0, NULL, ARRAY_SIZE(vk_image_barriers),
+                vk_image_barriers));
+
+        VK_CALL(vkCmdBeginRenderPass(list->vk_command_buffer, &begin_info, VK_SUBPASS_CONTENTS_INLINE));
+        VK_CALL(vkCmdBindPipeline(list->vk_command_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_info.vk_pipeline));
+        VK_CALL(vkCmdSetViewport(list->vk_command_buffer, 0, 1, &viewport));
+        VK_CALL(vkCmdSetScissor(list->vk_command_buffer, 0, 1, &scissor));
+        VK_CALL(vkCmdBindDescriptorSets(list->vk_command_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS,
+                pipeline_info.vk_pipeline_layout, 0, 1, &vk_descriptor_set, 0, NULL));
+        VK_CALL(vkCmdPushConstants(list->vk_command_buffer, pipeline_info.vk_pipeline_layout,
+                VK_SHADER_STAGE_FRAGMENT_BIT, 0, sizeof(push_args), &push_args));
+        VK_CALL(vkCmdDraw(list->vk_command_buffer, 3, region->dstSubresource.layerCount, 0, 0));
+        VK_CALL(vkCmdEndRenderPass(list->vk_command_buffer));
+
+        vk_image_barriers[0].oldLayout = attachment_layout;
+        vk_image_barriers[0].newLayout = dst_layout;
+        vk_image_barriers[0].srcAccessMask = dst_is_depth_stencil
+                ? VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT
+                : VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;
+
+        vk_image_barriers[1].oldLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+        vk_image_barriers[1].newLayout = src_layout;
+
+        VK_CALL(vkCmdPipelineBarrier(list->vk_command_buffer,
+                VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT |
+                VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT |
+                VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT |
+                VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT,
+                VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
```

Comment:
Should probably be TRANSFER_BIT as well.

Reply:
Is that still guaranteed to make the layout transition visible to whatever barrier the application might issue?

Reply:
The app will have to transition away from COPY_DEST state, in which case we get a transitive barrier.

Reply:
We block TRANSFER, and when app transitions away from COPY_DEST, it waits on TRANSFER, and we complete the barrier that way.

Reply:
Should be fixed.